pkg/webui/lib/components/with-locale.js

```diff
@@ -12,193 +12,179 @@
-    this.setState(state => ({ xx: !state.xx }))
-  }
 
-  componentDidUpdate(props) {
-    this.check(this.props, props)
-  }
-
-  componentDidMount() {
-    this.check({ user: {} }, this.props, true)
+const SUPPORTED_LOCALES = process.predefined.SUPPORTED_LOCALES // Note: defined by webpack define plugin.
+const defaultLanguage = 'en'
+const defaultEnglishRegion = 'US'
+const defaultLocale = `en-${defaultEnglishRegion}`
 
-    if (dev) {
-      window.addEventListener('keydown', this.onKeydown)
-      log('Press alt + L to toggle the xx locale')
-    }
-  }
-
-  @bind
-  check(prev, props, enforce = false) {
-    const current = (prev.user && prev.user.language) || envLocale
-    const next = (props.user && props.user.language) || envLocale || defaultLanguage
-
-    if (enforce || current !== next) {
-      this.promise = this.load(next)
-    }
-  }
-
-  @bind
-  success(p, withLocale) {
-    const newState = { loaded: true }
-    if (withLocale) {
-      const frontendMessages = p[0]
-      const backendMessages = p[1]
-
-      newState.messages = { ...frontendMessages, ...backendMessages }
-    }
-
-    this.setState(newState)
-  }
+const getLanguageName = (languageCode, type = 'language') => {
+  const languageNames = new Intl.DisplayNames(languageCode, { type })
+  return languageNames.of(languageCode)
+}
 
-  @bind
-  fail(err) {
+const supportedLocalesMap = SUPPORTED_LOCALES.reduce((acc, locale) => {
+  const loc = locale === 'en' ? `${locale}-${defaultEnglishRegion}` : locale
+  const language = loc.split('-')[0]
+  const region = loc.split('-').length > 1 ? loc.split('-')[1] : undefined
+
+  const title = region
+    ? `${getLanguageName(language)} (${getLanguageName(region, 'region')})`
+    : getLanguageName(language)
+
+  acc[loc] = title
+  return acc
+}, {})
+
+const m = defineMessages({
+  switchingLanguage: 'Switching language…',
+})
+
+export const LanguageContext = createContext()
+
+// `UserLocale` is a component that fetches the user's preferred language and
+// sets the language in the react-intl provider context. It will asynchronously
+// fetch translated messages and polyfills `window.Intl`.
+// The default language will be fetched from the env.
+const UserLocale = ({ children }) => {
+  const promise = useRef()
+  const [locale, setLocale] = useState(
+    localStorage.getItem('lang') ||
+      navigator.language ||
+      navigator.browserLanguage ||
+      defaultLanguage,
+  )
+  const [loaded, setLoaded] = useState(false)
+  const [messages, setMessages] = useState()
+
+  const handleIntlError = useCallback(err => {
     error(err)
-    this.setState({ messages: null, loaded: true })
-  }
-
-  @bind
-  onKeydown(evt) {
-    if (evt.altKey && evt.code === 'KeyL') {
-      this.toggle()
-    }
-  }
-
-  @bind
-  async load(language) {
-    let locale = navigator.language || navigator.browserLanguage || defaultLocale
+  }, [])
+
+  const load = useCallback(
+    async (locale, storeAsPreference = false) => {
```

Comment:
```suggestion
    async (locale) => {
```

`storeAsPreference` is not used anywhere